Updating forward declarations to reflect newer versions of the ed25519 APIs.

nmahendru opened the pull request:
#### c40282d7bbd5a49c66b9e67fb0eeed33bc0ed093
<pre>
Updating forward declarations to reflect newer versions of the ed25519 APIs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=262663">https://bugs.webkit.org/show_bug.cgi?id=262663</a>
<a href="https://rdar.apple.com/116490690">rdar://116490690</a>

Reviewed by Alex Christensen.

The interfaces were changed for Sonoma so the forward declarations need to be updated
to reflect the changes. Still keeping old changes for pre Sonoma builds.
This PR adds on to <a href="https://github.com/WebKit/WebKit/pull/19899">https://github.com/WebKit/WebKit/pull/19899</a> from @pascoej

Also, the newer APIs from corecrypto now return errors so calls to the same
have been updated with checks on return values.

* LayoutTests/platform/ios-16/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
* Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Cocoa.cpp:
(WebCore::signEd25519):
* Source/WebCore/crypto/mac/CryptoAlgorithmX25519Cocoa.cpp:
(WebCore::CryptoAlgorithmX25519::platformDeriveBits):
* Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):

Canonical link: <a href="https://commits.webkit.org/272946@main">https://commits.webkit.org/272946@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a00b449ff5f8e9183da8fa248f9cb3912c2a15fa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12416 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35571 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36260 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30518 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34688 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14789 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9573 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29609 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34118 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10468 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29997 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9142 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9230 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30005 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37586 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/28668 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30536 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30327 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35369 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/33598 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9373 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7326 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33254 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11151 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/40111 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7784 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9957 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8400 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10145 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on LayoutTests/platform/ios-16/TestExpectations:
```diff
@@ -23,11 +23,11 @@ media/media-webm-opus-variable-length.html [ Failure ]
 media/media-source/media-managedmse-resume-after-stall.html [ Skip ]
 
 # webkit.org/b/262663 (REGRESSION(iOS17/Sonoma): 8 WebCryptoAPI tests are constantly crashing.)
```
We should be able to delete this whole directory now.

nmahendru replied:
Let me follow up with a separate PR for that. So that we can revert/investigate easily if we need to later.

nmahendru replied:
https://github.com/WebKit/WebKit/pull/22557

aproskuryakov commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1623,6 +1623,12 @@
 
+#if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 140000) \
+    || (PLATFORM(MACCATALYST) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 170000) \
+    || (PLATFORM(IOS) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 170000)
```
No need for iOS version check, we don't support anything older.

nmahendru replied:
removed.

aproskuryakov commented on Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
```diff
@@ -65,6 +65,17 @@ const struct ccdigest_info *ccsha512_di(void);
 extern const struct ccdigest_info ccsha512_ltc_di;
 
 void cccurve25519(ccec25519key out, const ccec25519secretkey sk, const ccec25519base);
```
This function has also changed, why do we not need to update it too?

nmahendru replied:
right. I can add that too. Probably also need to check it's return value now and handle that.

nmahendru replied:
Have included other functions as well now and am now checking their return values.

aproskuryakov reviewed (commented):
This is probably fine to land as long as it fixes the tests, however I don't quite understand the big picture. If we are redefining the functions anyway, why do we care about matching SPI declarations?

sysrqb commented on Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
```diff
@@ -65,6 +65,17 @@ const struct ccdigest_info *ccsha512_di(void);
+#if HAVE(UPDATED_CORE_CRYPTO_SIGNATURES)
+
+int cccurve25519_make_priv(struct ccrng_state *, ccec25519secretkey);
+
```
nit: no need for separating all of these with spaces

nmahendru replied:
removed.

sysrqb commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1623,6 +1623,12 @@
+#if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 140000) \
+    || (PLATFORM(MACCATALYST) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 170000) \
+    || (PLATFORM(IOS) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 170000)
+#define HAVE_UPDATED_CORE_CRYPTO_SIGNATURES 1
```
Maybe we should use a slightly more descriptive name than "updated". Is there a CoreCrypto version number or scope this `#define` to curve/ed25519 signatures?

aproskuryakov replied:
`_WITH_INT_RETURN_VALUE`?

nmahendru replied:
CoreCrypto does have version numbers(just confirmed).
`_WITH_INT_RETURN_VALUE` seems better.. I can also add the commit hash of CC for which we are adding this ifdef.
How about `UPDATED_CORE_CRYPTO_SIGNATURES_WITH_INT_RETURN_VALUES_d0f475e405` ?

aproskuryakov replied:
I don't see the reason for adding commit hash.

achristensen07 commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1623,6 +1623,11 @@
 #endif
 
+#if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 140000) \
+    || (PLATFORM(MACCATALYST) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 170000)
```
|| PLATFORM(MACCATALYST) || PLATFORM(IOS_FAMILY)

nmahendru replied:
done.

nmahendru committed:
Updating forward declarations to reflect newer versions of the ed25519 APIs.
https://bugs.webkit.org/show_bug.cgi?id=262663
rdar://116490690

Reviewed by Alex Christensen.

The interfaces were changed for Sonoma so the forward declarations need to be updated
to reflect the changes. Still keeping old changes for pre Sonoma builds.
This PR adds on to https://github.com/WebKit/WebKit/pull/19899 from @pascoej

Also, the newer APIs from corecrypto now return errors so calls to the same
have been updated with checks on return values.

* LayoutTests/platform/ios-16/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/cocoa/CoreCryptoSPI.h:
* Source/WebCore/crypto/mac/CryptoAlgorithmEd25519Cocoa.cpp:
(WebCore::signEd25519):
* Source/WebCore/crypto/mac/CryptoAlgorithmX25519Cocoa.cpp:
(WebCore::CryptoAlgorithmX25519::platformDeriveBits):
* Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):

Canonical link: https://commits.webkit.org/272946@main
